### PR TITLE
Migrating the Gradient Transfer Notebook to use LLaMA-2

### PR DESCRIPTION
**Pull request by emersodb**
This notebook is investigating whether gradient optimized prompts for T5 transfer to improve the performance of LLaMA-2 on the SST2 task. The prompts were optimized using AutoPrompt in previous code that was available in the old PE Lab. However, I think the example is still informative even if participants cannot optimize the prompts themselves. 

That short answer is that these prompts **do not transfer**. We can see that they improve the Hugging Face T5 model, but not LLaMA-2.

### CODE REVIEW

File: src/reference_implementations/prompting_vector_llms/README.MD [hunk 1]
```diff
@@ -88,4 +88,4 @@ One of the most effective ways to help LLMs perform such tasks is through reason
...
+In the folder `transferring_gradient_optimized_prompts/`, we consider whether the gradient optimized prompts that we have found during discrete prompt optimization through gradient based search for the T5 model are transferrable to the LLaMA-2 7B model for sentiment analysis. This model is larger than T5, but has a different tokenizer and a different architecture. T5 has an encoder-decoder coupling, while Falcon is decoder-only. The prompts were optimized using AutoPrompt with a Vector implementation [here](https://github.com/VectorInstitute/PromptEngineering/blob/main/src/reference_implementations/prompt_zoo/gradient_search.py). Due to the compute intensive nature and complexity of the code, the implementation source code has not been included in the lab.
```
Comment (yuchongzhang) [inline src/reference_implementations/prompting_vector_llms/README.MD]: You mean "LLaM2-2 is decoder-only" instead of Falcon perhaps?

Comment (emersodb) [reply src/reference_implementations/prompting_vector_llms/README.MD]: Yes! I was originally going to use falcon, but forgot I needed activations.

File: src/reference_implementations/prompting_vector_llms/transferring_gradient_optimized_prompts/transferring_optimized_prompts.ipynb [hunk 2]
```diff
@@ -91,50 +198,81 @@
...
-    "dataloader = DataLoader(dataset, batch_size=10, shuffle=True)\n",
+    "dataloader = DataLoader(dataset, batch_size=batch_size, shuffle=True)\n",
     "dataset[0]"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In order to facilitate answer extraction, we need to know the tokenization ids for each of the labels we expect the model to response with. In this case, we optimized the prompts for T5 to produce either \"negative\" or \"positive.\" So we use the tokenizers to extract the IDs associated with these labels."
```
Comment (yuchongzhang) [inline src/reference_implementations/prompting_vector_llms/transferring_gradient_optimized_prompts/transferring_optimized_prompts.ipynb]: response with -> respond with

Comment (emersodb) [reply src/reference_implementations/prompting_vector_llms/transferring_gradient_optimized_prompts/transferring_optimized_prompts.ipynb]: Fixed 🙂

File: src/reference_implementations/prompting_vector_llms/transferring_gradient_optimized_prompts/transferring_optimized_prompts.ipynb [hunk 3]
```diff
@@ -91,50 +198,81 @@
...
      ]
     }
    ],
    "source": [
-    "# Need to grab the token id associated with the label words for both opt and t5\n",
-    "\n",
     "dummy_sentence = \"I love this movie!\"\n",
-    "print(opt_tokenizer(f\"{initial_prompt}{dummy_sentence} negative\")[\"input_ids\"])\n",
-    "print(opt_tokenizer(f\"{initial_prompt}{dummy_sentence} positive\")[\"input_ids\"])\n",
+    "print(llama_tokenizer(f\"{initial_prompt}{dummy_sentence} negative\")[\"input_ids\"])\n",
```
Comment (yuchongzhang) [inline src/reference_implementations/prompting_vector_llms/transferring_gradient_optimized_prompts/transferring_optimized_prompts.ipynb:239]: Just curious: is there a reason why we needed to use the dummy sentence and initial prompt here instead of merely providing the label words as inputs to the tokenizer?

Comment (emersodb) [reply src/reference_implementations/prompting_vector_llms/transferring_gradient_optimized_prompts/transferring_optimized_prompts.ipynb:239]: That's a good question. Depending on the tokenizer, the ID of a word can be different if it's on its own vs. as part of a sentence. For example, having a space in front of it or not. Since we're using it at the end of a prompt, we consider the token ID as part of a sentence to make sure we get the right one.

Comment (emersodb) [reply src/reference_implementations/prompting_vector_llms/transferring_gradient_optimized_prompts/transferring_optimized_prompts.ipynb:239]: I'll put a comment to that affect in the notebook for clarity.

Comment (yuchongzhang) [reply src/reference_implementations/prompting_vector_llms/transferring_gradient_optimized_prompts/transferring_optimized_prompts.ipynb:239]: Ah yes that totally makes sense!

File: src/reference_implementations/prompting_vector_llms/transferring_gradient_optimized_prompts/transferring_optimized_prompts.ipynb [hunk 4]
```diff
@@ -578,7 +723,7 @@
     "    # compute per-token log probability in a sequence.\n",
     "    # log_p has log probabilities for the following target output: [pos, it, ive]\n",
```
Comment (yuchongzhang) [inline src/reference_implementations/prompting_vector_llms/transferring_gradient_optimized_prompts/transferring_optimized_prompts.ipynb]: I'm a little confused about this comment. Are we saying that the tokenizer has split the label word "positive" into three tokens [pos, it, ive], and we have to sum up the log probabilities for each of them to obtain the log probability for the label word? I assume we are treating the label word as a sequence here?

File: src/reference_implementations/prompting_vector_llms/transferring_gradient_optimized_prompts/transferring_optimized_prompts.ipynb [hunk 5]
```diff
@@ -658,7 +810,7 @@
    "source": [
     "correct = 0\n",
     "total = 0\n",
-    "# We're going to use a loss function to extra the log probabilties of the labels.\n",
+    "# We're going to use a loss function to extra the log probabilities of the labels.\n",
```
Comment (yuchongzhang) [inline src/reference_implementations/prompting_vector_llms/transferring_gradient_optimized_prompts/transferring_optimized_prompts.ipynb]: "extra" -> "extract" perhahps?

**Review by yuchongzhang — APPROVED** [review]
Thanks for answering my questions. Everything looks good to me!
